src/table.rs

```diff
@@ -259,9 +286,194 @@ where
+            .collect()
+    }
+
+    /// Autoregressively sample n_samples of k characters from a kneser-ney smoothed n-gram model
+    pub fn kn_batch_sample(
+        &self,
+        query: &[u16],
+        n: usize,
+        k: usize,
+        n_samples: usize,
+        vocab: Option<u16>
+    ) -> Result<Vec<Vec<u16>>> {
+        (0..n_samples)
+            .into_par_iter()
+            .map(|_| self.kn_sample(query, n, k, vocab))
+            .collect()
+    }
+
+    /// Autoregressively sample k characters from a kneser-ney smoothed n-gram model
+    pub fn kn_sample(
+        &self,
+        query: &[u16],
+        n: usize,
+        k: usize,
+        vocab: Option<u16>,
+    ) -> Result<Vec<u16>> {
+        let mut rng = thread_rng();
+        let mut sequence = Vec::from(query);
+
+        for _ in 0..k {
+            let start = sequence.len().saturating_sub(n - 1);
+            let prev = &sequence[start..];
+            let probs = self.kn_probs(prev, vocab);
+
+            let dist = WeightedIndex::new(&probs)?;
+            let sampled_index = dist.sample(&mut rng);
+
+            sequence.push(sampled_index as u16);
+        }
+
+        Ok(sequence)
+    }
+
+    // Map of frequency to number of unique n-grams that occur with that frequency
+    fn compute_ngram_counts(&self, n: usize) -> HashMap<usize, usize> {
+        let mut count_map: HashMap<usize, usize> = HashMap::new();
+        
+        let query = vec![0; 0];
+        let (range_start, range_end) = self.boundaries(&query);
+        let mut stack = vec![(range_start, range_end, 1, query)];
+        'outer: while let Some((search_start, search_end, level_n, query)) = stack.pop() {
+            if search_start == search_end {
+                continue;
+            }
+
+            // Find median index of suffixes starting with `query` and ending with at least one additional token
+            let mut idx = search_start + (search_end - search_start) / 2;
+            while self.suffix(idx).eq(&query) {
+                idx = idx + (search_end - idx) / 2 + 1;
+                if idx >= search_end {
+                    continue 'outer;
+                }
+            }
+            
+            let token = self.suffix(idx)[query.len()];
+            let mut query_vec = query.clone();
+            query_vec.push(token);
+
+            let (start, end) = self.range_boundaries(&query_vec, search_start, search_end);
+            if level_n < n {
+                stack.push((start, end, level_n + 1, query_vec));
+            } else {
+                *count_map.entry(end - start).or_insert(0) += 1;
+            }
+
+            // Count other completions
+            if search_start < start {
+                stack.push((search_start, start, level_n, query.clone()));
+            }
+            if end < search_end {
+                stack.push((end, search_end, level_n, query));
+            }
+        }
+        count_map
+    }
+
+    fn get_delta(&self, n: usize) -> f64 {
```

Comment:
As in page 16 of https://people.eecs.berkeley.edu/~klein/cs294-5/chen_goodman.pdf